virt: allow defining the VM type and arch when creating it

cbosdo opened the pull request:
### What does this PR do?

Some hypervisors can handle several CPU architectures or have different
virtualization types. This is reflected in libvirt by the OS type (badly
named, indeed) and the arch value. Allow users to set them when creating
a VM using either virt.init or virt.running.

### What issues does this PR fix or reference?

None

### Previous Behavior

`virt.init` and `virt.running` where forcing x86_64 HVM virtual machines when creating them.

### New Behavior

User can choose the architecture and VM type to be created

### Tests written?

Yes

### Commits signed with GPG?

Yes

cbosdo committed:
virt: allow defining the VM type and arch when creating it

Some hypervisors can handle several CPU architectures or have different
virtualization types. This is reflected in libvirt by the OS type (badly
named, indeed) and the arch value. Allow users to set them when creating
a VM using either virt.init or virt.running.

Signed-off-by: Cédric Bosdonnat <cbosdonnat@suse.com>

rallytime commented on salt/modules/virt.py:
```diff
@@ -1224,6 +1223,16 @@ def init(name,
+        type of virtualization as found in the ``//os/type`` element of the libvirt definition.
+        The default value is taken from the host capabilities, with a preference for ``hvm``.
+
+        .. versionadded:: Fluorine
```
This should be `Neon` now.

rallytime commented on salt/modules/virt.py:
```diff
@@ -1224,6 +1223,16 @@ def init(name,
+        architecture of the virtual machine. The default value is taken from the host capabilities,
+        but ``x86_64`` is prefed over ``i686``.
+
+        .. versionadded:: Fluorine
```
`Neon` here as well. :)

rallytime commented on salt/modules/virt.py:
```diff
@@ -1200,6 +1197,8 @@ def _get_merged_nics(hypervisor, profile, interfaces=None, dmac=None):
 def init(name,
          cpu,
          mem,
+         os_type=None,
```
These new parameters need to go at the end of the list so we don't break things for people if they are using positional arguments.

rallytime commented on salt/states/virt.py:
```diff
@@ -239,6 +239,8 @@ def powered_off(name, connection=None, username=None, password=None):
 def running(name,
             cpu=None,
             mem=None,
+            os_type=None,
```
Same comments in this file as in the module above: The arguments need to be moved down and the versionadded tags should be `Neon` instead of Fluorine.

cbosdo replied:
Fixed them all.

rallytime reviewed (changes requested):
Thanks @cbosdo! I have some small requests.

cbosdo committed:
virt: disk target device depends on the bus

Not all disk target devices should start with 'vd' for non-vmware
hypervisors. Map the bus type to the right device name prefix. While at
it, also make sure the disks are created right for xen.

cbosdo commented:
The second commit being a bug fix, may be backported to Fluorine.

meaksh commented:
Hey @rallytime , the changes on this PR are part of @cbosdo's work around the virt module which has been already included in `fluorine`.

Some of the related PRs are https://github.com/saltstack/salt/pull/48736, https://github.com/saltstack/salt/pull/48421, https://github.com/saltstack/salt/pull/48261, https://github.com/saltstack/salt/pull/48262, etc. which are all included on `fluorine`.

Since this PR is still related with the previous ones and also contains a bugfix, it would be really nice to get this PR also into `fluorine`.

Would that be even possible at this point in time? Thanks!

cachedout commented:
@meaksh and @cbosdo Do you anticipate the need for additional backports beyond this or would this be the only one?

cbosdo commented:
@cachedout I don't expect any other backports: the other ones can wait for a future release

rallytime commented:
@cbosdo That sounds good. In the future let's keep the feature additions and bug fixes in separate PRs so it's easier to get the bug fixes in the right place. I am OK with backporting this one, but just in the future to make it simpler. :D

cbosdo commented:
@rallytime noted

rallytime commented:
Thanks! I'll get this backported right now.

meaksh commented:
Thanks @rallytime ! :+1:

rallytime commented:
@cbosdo Actually, this isn't backporting cleanly and is including more changed files than it should because of some additional work it's pulling in from a kwarg addition around the `loader`.  Can you make this change against the `fluorine` branch directly instead of me back-porting it?

cbosdo commented:
@rallytime sure

cbosdo commented:
@rallytime Should I change the versionadded back to Fluorine, then?

rallytime commented:
@cbosdo Yeah! Thanks for catching that.

rallytime commented:
@cbosdo No need to do anything for develop. I will merge forward and there will be a conflict on those lines. I'll take the `Fluorine` version. Thank you for following up!
